### PR TITLE
Added static tagcloud plugin.

### PR DESCRIPTION
**Pull request by felixfontein**
Allows to create HTML fragments and CSS for tag clouds which can be included in the blog via JavaScript or some post-processing.

### CODE REVIEW

File: v7/static_tagcloud/conf.py.sample [hunk 1]
```diff
@@ -0,0 +1,90 @@
+# Renders two tag clouds for tags and one for categories:
+# - a small one wiht at most 40 tags;
```
Comment (Kwpolska) [inline v7/static_tagcloud/conf.py.sample]: with*

File: v7/static_tagcloud/static_tagcloud/__init__.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+        # Compose list of tags, their post count and links
+        def acceptor(post):
+            return True if self.site.config['SHOW_UNTRANSLATED_POSTS'] else post.is_translation_available(lang)
+
+        tag_count_url_list = []
+        for tag in natsort.humansorted(list(posts_per_tag.keys())):
+            tag_count_url_list.append((
+                taxonomy.get_classification_friendly_name(tag, lang),
+                len([post for post in posts_per_tag[tag] if acceptor(post)]),
```
Comment (Kwpolska) [inline v7/static_tagcloud/static_tagcloud/__init__.py]: you can easily replace `acceptor` with the inlined logic, which can be simplified to
```py
self.site.config['SHOW_UNTRANSLATED_POSTS'] or post.is_translation_available(lang)
```

File: v7/static_tagcloud/conf.py.sample [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+RENDER_STATIC_TAGCLOUDS = {
+    'tag-small': {
+        # Tag cloud's name (used as CSS class). {0} will be replaced
+        # by the language.
+        'name': 'tcs-{0}',
+
+        # Filename for the HTML fragment. {0} will be replaced by the
+        # language.
+        'filename': 'tagcloud-{0}.inc',
```
Comment (Kwpolska) [inline v7/static_tagcloud/conf.py.sample]: `.inc.html` looks better to me (but feel free to ignore that)

File: v7/static_tagcloud/static_tagcloud/engine.py [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+        if tag_weight in tag_weight_to_level_weight:
+            level_weight, level_index = tag_weight_to_level_weight[tag_weight]
+            for name, url, index in list:
+                tags.append((index, name, level_weight, level_index, url))
+    tags.sort()
+    return tags, level_weights
+
+
+def _get_hex_color(color):
+    def myHex(i, digits):
```
Comment (Kwpolska) [inline v7/static_tagcloud/static_tagcloud/engine.py]: String formatting would work better for that, and `str.join` can take a generator expression:

```py
def _get_hex_color(color):
    return '#' + ''.join('{:02x}'.format(int(round(c * 255))) for c in color)
```

File: v7/static_tagcloud/conf.py.sample (hunk 1 above)
Comment (felixfontein) [inline v7/static_tagcloud/conf.py.sample]: thanks!

File: v7/static_tagcloud/conf.py.sample (hunk 3 above)
Comment (felixfontein) [inline v7/static_tagcloud/conf.py.sample]: Changed that.

File: v7/static_tagcloud/static_tagcloud/engine.py (hunk 4 above)
Comment (felixfontein) [inline v7/static_tagcloud/static_tagcloud/engine.py]: Good idea.

File: v7/static_tagcloud/static_tagcloud/__init__.py (hunk 2 above)
Comment (felixfontein) [inline v7/static_tagcloud/static_tagcloud/__init__.py]: Done.

**Comment by Kwpolska** [comment]
Merge at will, the site will (or at least, should) update itself on midnight UTC.

**Comment by felixfontein** [comment]
Ok, thanks!

**Comment by felixfontein** [comment]
@Kwpolska: it looks like the plugin site didn't update (correctly).

**Comment by Kwpolska** [comment]
What went wrong?

**Comment by felixfontein** [comment]
The page for sidebar (https://plugins.getnikola.com/#sidebar) still has the broken ReST-style link, and the page for static_tagcloud doesn't seem to exist (https://plugins.getnikola.com/#static_tagcloud).

**Comment by Kwpolska** [comment]
I can see it, force a reload (Ctrl+Shift+R) and blame our single-page architecture.

**Comment by felixfontein** [comment]
Now both vanished for me...

**Comment by Kwpolska** [comment]
Try refreshing [the JS file](https://plugins.getnikola.com/plugin_data.js) or using incognito mode?

**Comment by felixfontein** [comment]
It's really strange. If I load the JS file via `wget` or open it in a new tab, I can see `sidebar` and `static_tagcloud` in there. But when I check out via inspector what the browser is using on the pages, it uses an old version. Forcing reload or using incognito mode in Chromium (I'm testing both Firefox and Chromium) doesn't help, the same thing is happening there. Also, clearing the cache has the same result, which I find even more strange!

**Comment by felixfontein** [comment]
Hmm, maybe this is all related to CloudFlare? They seem to convert JavaScript to "rocketscript", and their rocket scripts are doing the actual loading, so this could be completely independent from what the browser is doing.

**Comment by felixfontein** [comment]
I guess I get the content from another CloudFlare server than you do, and the one I get the content from serves an outdated version of `plugin_data.js`.

**Comment by Kwpolska** [comment]
I tried nuking that file from cache. I should probably get going on the JS-less version of that site.

**Comment by felixfontein** [comment]
https://github.com/Kwpolska/nikola-plugins-v2 is going to be the JS-less version, right?

**Comment by Kwpolska** [comment]
Yes.

**Comment by felixfontein** [comment]
Can I help with that?

**Comment by Kwpolska** [comment]
Sure, I just invited you to that project. I think there isn’t *that* much to do, and to run it, you need to copy `v7` from this repo into the root directory.

**Comment by felixfontein** [comment]
Where does it expects the plugin repository to be?

**Comment by felixfontein** [comment]
Ah. I tried that, but now I get an exception. Looks like one of the plugins is missing a `Nikola` section in the config.

**Comment by felixfontein** [comment]
Actually, that's `v7/notebook_shortcode/notebook_shortcode.plugin`.

**Comment by felixfontein** [comment]
(And is there a better place to discuss this then here?)

**Comment by Kwpolska** [comment]
getnikola/nikola#2482

I fixed `notebook_shortcode.plugin`.

**Comment by felixfontein** [comment]
Ok, thanks!

**Pull request merged**